Add Income placeholder page

sboisvert opened the pull request:
To test https://cra-alpha-pr-29.herokuapp.com/

To add the income placeholder page we need to change the data structure from the "API" and change how the dashboard uses the income data.

TODO: Need unit tests

sboisvert committed:
Initial commit of T4 page

sboisvert committed:
Add income array in API

sboisvert committed:
Make Dashboard Page follow to T4

Instead of going to the confirmation page let's go from the Dashboard to the income page. This flow will change when we add more placeholders.

sboisvert committed:
Add T4 route

sboisvert committed:
Try to get income data from "API"

This isn't working yet. Not sure what syntax I'm supposed to use.

sboisvert committed:
Update sample "API" values

Change to an object to make dereferencing easier.

sboisvert committed:
Clean up console.log and small style changes

sboisvert committed:
Merge branch 'master' into t4

sboisvert committed:
Update Dashboard with new income structures

pcraig3 commented on src/pages/Dashboard.js:
```diff
@@ -31,7 +31,13 @@ const yourFamilyRows = ({ maritalStatus, children }) => {
 }
 
 const yourIncomeRows = ({ income }) => {
```
So you _could_ do (since you know everything is nested inside of the "income" key). 
I would generally prefer that, but this data is pretty changeable.

```suggestion
const yourIncomeRows = ({ income: { employerName, year, box12, box14, box22 } = {} }) => { 
return [
    { key: 'Employer Name', value: employerName },
    { key: 'Year', value: year },
    ...
  ]
  }
```

sboisvert replied:
I agree that I like this style better. Is the `= {}` a default value that's passed to the function?

pcraig3 replied:
Yeah, if there's no value passed in (ie, you call `yourIncomeRows()`), then you default `income` to an empty object.

Otherwise, `income` will default to `undefined`, and trying to access `income.employerName` will cause runtime errors.

pcraig3 commented on src/pages/T4.js:
```diff
@@ -0,0 +1,60 @@
+
+const changesButtonCSS = css`
+  width: 200px;
+  background-color: pink;
```
In my [About You PR](https://github.com/cds-snc/cra-alpha/pull/28/files#diff-65c1b1a621d382085d912a379ad71c9b), I pulled this styling out, but we can refactor this later. So no worries.

pcraig3 commented on src/pages/T4.js:
```diff
@@ -0,0 +1,60 @@
+`
+
+//todo How to get t4 info? Where does it get injected from? user object passed with an array of t4s?
+//is this passed in to the Dashbord object which then passes it to summary table?
```
All good questions.

Currently it's just a user object that goes into the session.

pcraig3 commented on src/pages/T4.js:
```diff
@@ -0,0 +1,60 @@
+    <${Layout}>
+      <div class=${t4CSS}>
+        <${LogoutLink} />
+        <h1>Income</h1>
```
Let's call this "Your income" to keep things about you 👉 😄

pcraig3 commented on src/pages/T4.js:
```diff
@@ -0,0 +1,60 @@
+        <${LogoutLink} />
+        <h1>Income</h1>
+        <div>
+        <img src='/t4.png' />
```
I know it's just a demo image, but can you put an `alt` tag in, maybe "sample T4 form"? I'm trying to keep it as accessible as possible.

I've been running all my pages through the [axe chrome extension](https://chrome.google.com/webstore/detail/axe/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US), just in the browser. (There's probably a Firefox equivalent). 

<img width="1186" alt="Screen Shot 2019-04-18 at 3 06 33 PM" src="https://user-images.githubusercontent.com/2454380/56384945-bb5aa700-61eb-11e9-800f-1643a9bd1095.png">

pcraig3 commented on src/api.js:
```diff
@@ -5,7 +5,13 @@ var API = (function() {
     children: '8',
-    income: '$10,000',
+    income: {
+      employerName: 'Acme Corp',
```
Since we're adding a company name now, let's add this to the intro paragraph on the main dashboard page.

(Also, v sad to lose BLORB CORP 😢 )

pcraig3 commented on src/api.js:
```diff
@@ -23,7 +35,13 @@ var API = (function() {
+      employerName: 'Aplha Corp',
+      year: 2019,
+      box12: '254528961',
+      box14: 50000,
```
I know it's a small thing, but I think we should keep the income less than 40000 since that's supposed to be our threshold.

sboisvert replied:
Nod, in the future we might want to have those to do our error checking or something, as in, do you have too much income for this tool.

pcraig3 commented on src/pages/T4.js:
```diff
@@ -0,0 +1,60 @@
+
+<${SummaryTable} title="Income Data" rows=${t4Data(data.income)} ifEditable=${false} //>
+<${ButtonLink} href="/confirmation" style=${changesButtonCSS}>I need to make changes<//>
+<${ButtonLink} href="/confirmation" style=${consentButtonCSS}>This information is accurate<//>
```
The GOV.UK button guidance (https://design-system.service.gov.uk/components/button/#how-it-works) usually has succinct button text, and it also says "Align the primary action button to the left edge of your form." 

So let's keep the "positive" action on the left hand side.

pcraig3 commented on src/server.js:
```diff
@@ -39,6 +39,17 @@ app.get('/', (req, res) => {
+    renderPage({
+      locale,
+      pageComponent: 'T4',
+      title: 'T4',
```
Let's keep the page title the same as the H1.

pcraig3 reviewed (changes requested):
A few small things, but nothing major.

I have an example unit test for a new page in my other PR, so you could probably just crib off that one.

https://github.com/cds-snc/cra-alpha/pull/28/files#diff-732cf9bd6ab74c59e970f189ad24a78b

pcraig3 commented on src/pages/T4.js:
```diff
@@ -0,0 +1,60 @@
+const ButtonLink = require('../components/ButtonLink.js')
+const ValidationError = require('../components/forms/ValidationError.js')
+const Checkbox = require('../components/forms/Checkbox.js')
+const Button = require('../components/forms/Button.js')
```
Also, I didn't notice earlier, but the linter is complaining about some unused imports. 

![image](https://user-images.githubusercontent.com/2454380/56386484-e0e9af80-61ef-11e9-8727-91185541482f.png)

(You can run `npm run lint`.)

sboisvert committed:
Update amounts

Since the service is only for people with lower incomes it would be best to have only that data in for now.

sboisvert committed:
Remove needlessly imported dependencies

sboisvert committed:
Change order of buttons

sboisvert committed:
Add default value to t4Data function declaration

sboisvert committed:
Change Income to "Your Income"

sboisvert committed:
Add title to image

sboisvert committed:
Update test data

Change how the data is tested to match how it is in the API

sboisvert committed:
Update Dashboard tests to correctly test for income

Now that we've added income as a sub-object we need to test correctly against it. Right now it's doing a nested check, but we may want to change to reccursion if we're going to have more than 1 level of nesting.

sboisvert committed:
Add T4 tests

Test for income properties showing up correctly on the T4 page.

sboisvert commented:
@pcraig3 I think I've resolved the issues

pcraig3 commented on src/pages/__tests__/Dashboard.test.js:
```diff
@@ -42,18 +42,35 @@ describe('<Dashboard>', () => {
-        ),
-      )
-      expect($('h1').text()).toEqual('Hi, Fred')
+    //If we have an object inside our data object (for example income info) we need to loop thru it and not just process the top most value
```
We could also make the top-level object into:
```
  {
    name: 'Fred Smith',
    address: 'Mississauga',
    maritalStatus: 'Married',
    children: 5,
    employerName: 'Blorb Corp',
    year: 2019,
    box12: '321987645',
    box14: 10000,
    box22: 1000,
  },
```

to keep our original loop, since we're still expecting the string to be on the page. (since we are testing the sub-object here: https://github.com/cds-snc/cra-alpha/pull/29/commits/c486b43d7c889f0f5a385c47ba8acb3a6c48dce4.)

But if we want to match the "user object" we have, then we can keep this as it is now.

sboisvert replied:
>if we want to match the "user object" we have, then we can keep this as it is now.

I think we want to match it as closely as possible so our test are as close to "real life" as possible.

pcraig3 commented on src/server.js:
```diff
@@ -39,6 +39,17 @@ app.get('/', (req, res) => {
+    renderPage({
+      locale,
+      pageComponent: 'T4',
+      title: 'T4',
```
Let's make the title `Your Income` so that it matches the `<h1>` on the page.

sboisvert replied:
oups, I thought I did that! Done now

pcraig3 commented:
![image](https://user-images.githubusercontent.com/2454380/56604409-da807c80-65cf-11e9-8efa-6e1b5f63553f.png)

Last thing is a linting error. You can run this test locally with `npm run dev` if it keeps catching you out.

sboisvert committed:
Update page title to match T4 page h1

sboisvert committed:
Fix var not used error

sboisvert committed:
fix up tests

Add an ID to the buttons so we can easily target them when we test them.

pcraig3 reviewed (approved):
🚢 🚢  looks good!
